wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGrpcGenerator.kt

```diff
@@ -0,0 +1,162 @@
+    servicePackageName: String?
+  ): FunSpec {
+    val packageName = if (servicePackageName.isNullOrBlank()) "" else "$servicePackageName."
+
+    val wireRpcAnnotationSpec = AnnotationSpec.builder(WireRpc::class.asClassName())
+        .addMember("path = %S", "/$packageName$serviceName/${rpc.name()}")
+        .addMember("requestAdapter = %S", "$packageName${rpc.requestType().simpleName()}#ADAPTER")
+        .addMember("responseAdapter = %S", "$packageName${rpc.responseType().simpleName()}#ADAPTER")
+        .build()
+    val funSpecBuilder = FunSpec.builder(rpc.name())
+        .addModifiers(KModifier.ABSTRACT)
+        .addAnnotation(wireRpcAnnotationSpec)
+        .addParameter("context", CoroutineContext::class.java)
+
+    when {
+      rpc.requestStreaming() && rpc.responseStreaming() -> {
+        val requestChannel =
+            SendChannel::class.asClassName().parameterizedBy(rpc.requestType().typeName)
+        val responseChannel =
+            ReceiveChannel::class.asClassName().parameterizedBy(rpc.responseType().typeName)
+        funSpecBuilder
+            .returns(Pair::class.asClassName().parameterizedBy(requestChannel, responseChannel))
+      }
+      rpc.requestStreaming() -> {
+        val requestChannel =
+            SendChannel::class.asClassName().parameterizedBy(rpc.requestType().typeName)
+        val responseDeferred =
+            Deferred::class.asClassName().parameterizedBy(rpc.responseType().typeName)
+        funSpecBuilder
+            .returns(Pair::class.asClassName().parameterizedBy(requestChannel, responseDeferred))
+      }
+      rpc.responseStreaming() -> {
+        funSpecBuilder
+            .addParameter("request", rpc.requestType().typeName)
+            .returns(
+                ReceiveChannel::class.asClassName().parameterizedBy(rpc.responseType().typeName))
+      }
+      else -> {
+        funSpecBuilder.addModifiers(KModifier.SUSPEND)
+            .addParameter("request", rpc.requestType().typeName)
+            .returns(rpc.responseType().typeName)
+      }
+    }
+
+    return funSpecBuilder.build()
+  }
+
+  companion object {
+    private val BUILT_IN_TYPES = mapOf(
+        ProtoType.BOOL to BOOLEAN,
+        ProtoType.BYTES to ByteString::class.asClassName(),
+        ProtoType.DOUBLE to DOUBLE,
+        ProtoType.FLOAT to FLOAT,
+        ProtoType.FIXED32 to INT,
+        ProtoType.FIXED64 to LONG,
+        ProtoType.INT32 to INT,
+        ProtoType.INT64 to LONG,
+        ProtoType.SFIXED32 to INT,
+        ProtoType.SFIXED64 to LONG,
+        ProtoType.SINT32 to INT,
+        ProtoType.SINT64 to LONG,
+        ProtoType.STRING to String::class.asClassName(),
+        ProtoType.UINT32 to INT,
+        ProtoType.UINT64 to LONG
+    )
+
+    @JvmStatic @JvmName("get")
+    operator fun invoke(
+      schema: Schema
+    ): KotlinGrpcGenerator {
+      val map = BUILT_IN_TYPES.toMutableMap()
+
+      fun putAll(kotlinPackage: String, enclosingClassName: ClassName?, types: List<Type>) {
```

Comment:
I wonder if there’s a way to avoid doing this twice

Reply:
will merge `KotlinGrpcGenerator` into `KotlinGenerator`